### PR TITLE
Replace special-purpose monotonic explode operators with `explode_one`

### PR DESCRIPTION
**Pull request by tokenrove**
We separate two ad-hoc operators into a monotonicity check and a call to `explode_one`, per #17606.  There are also two small cleanup commits included that didn't feel large enough to go in their own PRs.  This is closely related to, and modeled on, the work on #17178 and its design doc in #17597.

I'd like to have tests of the error paths but it's not clear how to easily provoke them.  I'm experimenting with this presently, and if I get time, will expand some the associated testdrive tests, but opening this for review now so as to avoid rabbitholing on that.

From a performance POV, it would be nice to demonstrate whether the benefits of `explode_one` outweigh the extra machinery for error handling.  This is also something I'm working on, but don't want to necessarily block this on.

Justification for `num_traits` dependency:
 - we already depend on this crate in other crates;
 - these traits only describe existing aspects of the standard library;
 - Signed feels like the least-specific specialization that allows the monotonicity check to be useful.

### Motivation

  * This PR fixes a recognized bug: #17606.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - No user-facing behavior changes, at least in the absence of erroneous sources or bugs.

### CODE REVIEW

**Comment by tokenrove** [comment]
Oops, I just noticed @vmarcos had said on the original issue:

> It is probably a good idea here to not only produce query-level errors, but also log errors to Sentry in the fallible monotonicity check operator. This behavior would be in line with the strategy being favored for epic https://github.com/MaterializeInc/materialize/issues/17178.

I'll look into adding the Sentry logging to this.

File: src/compute/src/render/top_k.rs
Comment (teskje) [inline src/compute/src/render/top_k.rs:337]: Could you add a newline here? I have the slight suspicion that long strings tend to break rustfmt. It sees them, decides that it can't do anything about them and just gets depressed and gives up.

**Review by teskje — APPROVED** [review]
LGTM, pending @vmarcos' approval. Just a couple nits!

File: src/compute/src/render/top_k.rs
Comment (tokenrove) [reply src/compute/src/render/top_k.rs:337]: I fully expected rustfmt to deal with this for me, woops.

File: src/compute/src/render/top_k.rs
Comment (antiguru) [inline src/compute/src/render/top_k.rs:336]: Same as above.

File: src/timely-util/src/operator.rs
Comment (antiguru) [reply src/timely-util/src/operator.rs:205]: > Maybe we should consider warn(missing_docs) in this crate.

Could you file an issue for that? I think it'd be nice to have!

**Review by antiguru — APPROVED** [review]
Thanks, this looks good. I left a comment about placing the check.

File: src/compute/src/render/top_k.rs
Comment (tokenrove) [reply src/compute/src/render/top_k.rs]: I was discussing with @vmarcos more generally this pattern of warn/error, and will file an issue around clarifying it and the exposure of customer data, once I've figured out the right phrasing, but it did strike me that we're probably going to do this a lot, and it might make sense to have a macro that has the
```
foo!(data=?data, ..., "message")
```
form since it could more easily do the `error!` without the data, and the `warn!` with.

File: src/compute/src/render/reduce.rs
Comment (antiguru) [reply src/compute/src/render/reduce.rs]: You might be able to select from one of the tables that have a long retention period. [`mz_cluster_replica_utilization`](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_cluster_replica_utilization) has a retention of 30 days, and reading from it with an as-of in the past would currently guarantee that the persist source replays a lot of history. That said, there might be easier ways to emulate this behavior, similar to the tricks @vmarcos was using to force invalid data into reductions for testing purposes.

Comment (tokenrove) [reply src/compute/src/render/reduce.rs]: Oh yeah, I had considered `repeat_row` as a strategy for producing bad input, but I hadn't thought about using it to create something that's good, but only once consolidated.  That's worth a shot.

File: src/timely-util/src/operator.rs
Comment (tokenrove) [reply src/timely-util/src/operator.rs:205]: Filed #18192; I had hoped that there were few enough existing warnings that I could just include that as a small change here, but there's juuust enough that it makes sense to do as a separate task.

File: src/compute/src/render/reduce.rs
Comment (vmarcos) [reply src/compute/src/render/reduce.rs]: Unfortunately, this is indeed a bit tricky to test currently. The problem is that `repeat_row` is marked to be monotonicity-breaking: https://github.com/MaterializeInc/materialize/blob/bae5ea5cfe28408bad1fe64e99a17b16803f099d/src/expr/src/relation/func.rs#L2510
It would be possible to do some manual testing by turning the above to `true`; not sure how I would feel about flipping it to `true` more definitely. It might be OK if we deem `repeat_row` a testing-only function intended to create incorrect behavior? Strictly, it is not monotonicity-preserving, though.
Another option is to add automated tests later when we work on #17358?

**Review by vmarcos — APPROVED** [review]
This LGTM! It is indeed a bit tricky to test these changes, as discussed in https://github.com/MaterializeInc/materialize/pull/18140#discussion_r1140140029

File: src/compute/src/render/reduce.rs
Comment (tokenrove) [reply src/compute/src/render/reduce.rs]: > Unfortunately, this is indeed a bit tricky to test currently. The problem is that `repeat_row` is marked to be monotonicity-breaking:

Alas, that explains the failure of my attempts to trigger this.  

> Another option is to add automated tests later when we work on #17358?

Agreed.  I think I will merge this for now, but continue thinking about the best way to approach testing all these error paths in general.

**Pull request merged**
